### PR TITLE
Document consensus rules from 4.6 Action Descriptions

### LINKED ISSUE #3218: Document consensus rules from Zcash spec: 4.6 Action Descriptions
For each consensus rule:

- check that the rule is implemented in Zebra
- make sure the latest version of the rule is quoted in the docs for the code that implements it
- make sure the code implements the latest version of the rule

## Consensus rules

https://zips.z.cash/protocol/protocol.pdf#actiondesc

Screenshot 2021-12-16 at 17 47 35

## Implementation
 

- #1864
    - PR: #1885

### LINKED ISSUE #3226: Document consensus rules from Zcash spec: 7.5 Action Description Encoding and Consensus
For each consensus rule:

- check that the rule is implemented in Zebra
- make sure the latest version of the rule is quoted in the docs for the code that implements it
- make sure the code implements the latest version of the rule

## Consensus rule

https://zips.z.cash/protocol/protocol.pdf#actionencodingandconsensus

Screenshot 2021-12-16 at 15 48 20

## Implementation
 
- #1864 (PR: #1885)

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

We must document all consensus rules and double-check they were implemented.

## Solution

Closes https://github.com/zcashfoundation/zebra/issues/3218
Also closes https://github.com/ZcashFoundation/zebra/issues/3226; that rule seems redundant with the first rule from 4.6.

## Review

@dconnolly and/or @upbqdn might want to review.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by conradoplg** [comment]
@Mergifyio update

File: zebra-chain/src/orchard/action.rs [hunk 1]
```diff
@@ -59,13 +59,47 @@ impl ZcashSerialize for Action {
...
+            // https://zips.z.cash/protocol/protocol.pdf#abstractcommit
+            // See [`ValueCommitment::zcash_deserialize`].
             cv: ValueCommitment::zcash_deserialize(&mut reader)?,
+            // Type is `{0 .. 𝑞_ℙ − 1}`. See [`Nullifier::try_from`].
             nullifier: Nullifier::try_from(reader.read_32_bytes()?)?,
+            // Type is SpendAuthSig^{Orchard}.Public, i.e. ℙ.
+            // https://zips.z.cash/protocol/protocol.pdf#concretespendauthsig
+            // https://zips.z.cash/protocol/protocol.pdf#concretereddsa
+            // This only reads the 32-byte buffer. The type is enforced
+            // on signature verification; see zebra-chain/src/primitives/redpallas/batch.rs
```
Comment (upbqdn) [inline zebra-chain/src/orchard/action.rs]: Wouldn't a reference like this work?
```suggestion
            // on signature verification; see [`redpallas::batch`]
```

File: zebra-consensus/src/transaction.rs [hunk 2]
```diff
@@ -922,9 +924,21 @@ where
...
+                // >   associated with an input, as defined in § 4.10 using SIGHASH_ALL.
+                // > - The spend authorization signature MUST be a valid SpendAuthSig^{Orchard}
+                // >   signature over SigHash using rk as the validating key — i.e.
+                // >   SpendAuthSig^{Orchard}.Validate_{rk}(SigHash, spendAuthSig) = 1.
+                // >   As specified in § 5.4.7, validation of the 𝑅 component of the
+                // >   signature prohibits non-canonical encodings.
+                //
+                // https://zips.z.cash/protocol/protocol.pdf#actiondesc
+                //
+                // This is validated by the verifier, inside the `redpallas` crate.
```
Comment (upbqdn) [inline zebra-consensus/src/transaction.rs]:
```suggestion
                // This is validated by the verifier, inside the `redpallas` module.
```

**Review by upbqdn — COMMENTED** [review]
Two suggestions, otherwise it looks good to me. I can approve once the conversations are resolved.

**Review by upbqdn — DISMISSED** [review]

File: zebra-consensus/src/transaction.rs [hunk 3]
```diff
@@ -922,9 +924,21 @@ where
...
+                // >   associated with an input, as defined in § 4.10 using SIGHASH_ALL.
+                // > - The spend authorization signature MUST be a valid SpendAuthSig^{Orchard}
+                // >   signature over SigHash using rk as the validating key — i.e.
+                // >   SpendAuthSig^{Orchard}.Validate_{rk}(SigHash, spendAuthSig) = 1.
+                // >   As specified in § 5.4.7, validation of the 𝑅 component of the
+                // >   signature prohibits non-canonical encodings.
+                //
+                // https://zips.z.cash/protocol/protocol.pdf#actiondesc
+                //
+                // This is validated by the verifier, inside the `redpallas` module.
```
Comment (upbqdn) [inline zebra-consensus/src/transaction.rs]: Actually, why don't we change it to a reference?
```suggestion
                // This is validated by the verifier, inside the [`primitives::redpallas`] module.
```

Comment (upbqdn) [reply zebra-consensus/src/transaction.rs]: It's also clearer where the reader should look for the `redpallas` module.

**Review by upbqdn — APPROVED** [review]

**Pull request merged**